[WPE] Memory-mapped GPU buffers fail under DRM platform

nikolaszimmermann opened the pull request:
#### e3ab4a2e6678b8bc55ef4a6c8ea6526ab83ffd80
<pre>
[WPE] Memory-mapped GPU buffers fail under DRM platform
<a href="https://bugs.webkit.org/show_bug.cgi?id=294995">https://bugs.webkit.org/show_bug.cgi?id=294995</a>

Reviewed by Carlos Garcia Campos.

When using the DRM platform on i.MX systems with separate GPU and IPU, memory-
mapped GPU buffers failed to initialize properly. The DRM implementation
correctly chooses the first KMS-capable device (e.g. /dev/dri/card1) for
display buffer allocation. However, it incorrectly propagates the same node
to DRMDeviceManager as &apos;drmRenderNode&apos; -- which is meant to be associated with
the &quot;main device&quot;. For an overview about the terminology (main device vs. target
device), see: <a href="https://wayland.app/protocols/linux-dmabuf-v1#zwp_linux_dmabuf_feedback_v1.">https://wayland.app/protocols/linux-dmabuf-v1#zwp_linux_dmabuf_feedback_v1.</a>

Memory-mapped GPU buffers are still able to allocate gbm_bos from card1,
and export them as dma-bufs -- but trying to mmap() will fail, with
EACCESS permission denied error. That operation is only possible for
buffers allocated using the &quot;main device&quot;.

The Wayland platform handles this correctly by listening to `main_device` and
`tranche_target_device&apos; linux-dmabuf-v1 events. The GPU is correctly used as
&apos;main device&apos;, whereas the display-controller is used as &apos;target device&apos; --
fix DRM to behave in a similar way. If the chosen DRM device has no associated
render node, then try to use the first DRM render node instead of not selecting
a render node.

Covered by existing tests, under DRM backend + linear tile textures
activated.

* Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
(findTargetDevice):
(findFirstRenderNodeName):
(wpeDisplayDRMSetup):
(findDevice): Deleted.

Canonical link: <a href="https://commits.webkit.org/297131@main">https://commits.webkit.org/297131@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/220f938f8a1b9b817a8d4be6155d2e4c8ba87ad5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110583 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30242 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20675 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116609 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60850 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30921 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38830 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84091 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113531 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24691 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99575 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64532 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24054 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60404 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94076 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17774 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119399 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37623 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27938 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93056 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37997 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95846 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92878 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23673 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37891 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15635 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33597 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37518 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42989 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37181 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40520 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38889 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/220f938f8a1b9b817a8d4be6155d2e4c8ba87ad5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47208

nikolaszimmermann committed:
[WPE] Memory-mapped GPU buffers fail under DRM platform
https://bugs.webkit.org/show_bug.cgi?id=294995

Reviewed by Carlos Garcia Campos.

When using the DRM platform on i.MX systems with separate GPU and IPU, memory-
mapped GPU buffers failed to initialize properly. The DRM implementation
correctly chooses the first KMS-capable device (e.g. /dev/dri/card1) for
display buffer allocation. However, it incorrectly propagates the same node
to DRMDeviceManager as 'drmRenderNode' -- which is meant to be associated with
the "main device". For an overview about the terminology (main device vs. target
device), see: https://wayland.app/protocols/linux-dmabuf-v1#zwp_linux_dmabuf_feedback_v1.

Memory-mapped GPU buffers are still able to allocate gbm_bos from card1,
and export them as dma-bufs -- but trying to mmap() will fail, with
EACCESS permission denied error. That operation is only possible for
buffers allocated using the "main device".

The Wayland platform handles this correctly by listening to `main_device` and
`tranche_target_device' linux-dmabuf-v1 events. The GPU is correctly used as
'main device', whereas the display-controller is used as 'target device' --
fix DRM to behave in a similar way. If the chosen DRM device has no associated
render node, then try to use the first DRM render node instead of not selecting
a render node.

Covered by existing tests, under DRM backend + linear tile textures
activated.

* Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
(findTargetDevice):
(findFirstRenderNodeName):
(wpeDisplayDRMSetup):
(findDevice): Deleted.

Canonical link: https://commits.webkit.org/297131@main
